### ISSUE TITLE
Astronomical Units

### ISSUE THREAD

**Issue by nwalet** [maths/moodle-qtype_stack#401]
It would be lovely to have astronomical distance units in the units packages:
1 AU (Astronomical Unit) = 1.49597871e11 m
1 pc (parsec) = 3.08567758e16 m
and maybe (but maybe not!) the ly/lightyear= 9.4605284e15 m

**Comment by sangwinc** [comment]
Niels,
Very happy to add these in these kinds of things if it helps you.  Can you confirm this is what you need?

Also, can I ask you to propose some non-trivial test cases around here:
https://github.com/maths/moodle-qtype_stack/blob/iss401/tests/fixtures/answertestfixtures.class.php#L1380

I'm a bit concerned as we overload the notation that we will get unintended clashes. I can't think of any here, but "km" is literally "kilo*meters".  I don't want "pc" to be "pico*?" as well, so we should put tests in now to fix the intended behaviour for the future.

**Comment by aharjula** [comment]
pc is pretty common and I don't think anyone would mix that with the prefix p. But lightyear is something we will have trouble as we currently define a "year" as we do and we do not have a matching constant for "c" to give that value at sufficient accuracy. In any case if we get parsec we should not define it as a raw number like "3.08567758e16 m" instead we should use its definition thus keeping our magic number count down to just defining the AU. And we really should have a common constants library that we could load on top of the units, all those physics [constants](https://physics.info/constants/) and stuff like AU and c would be nice to have.

**Comment by nwalet** [comment]
Yes, it is. Will add some tests as requested. I tried to be careful in my choice, which is why I probably don't want the lightyear--overloading too much is indeed dangerous. (AU is already bad as well since the automic unit au is used as a replacement for 1 Hartree=2Rydberg=27.2 eV ).

Looking at the unit sets, I would probably have designed slightly differently: I would have concentrated on core SI, adding other sets as required (i.e., if I want eV, astronomical, imperial, or any other such unit set, I need to set a switch in the question). That would add additional safety, and can make the same non-SI unit mean different things dependening on the context for the question--it would also mean that there is only limited overlaod.

**Comment by sangwinc** [comment]
I think you are right about the design here: we probably do want a more flexible "unit sets" system to add in optional units.

**Comment by sangwinc** [comment]
Niels,
Are you still interested in this issue?  If so, perhaps you could propose some units we need to add in.
